library/java/org/chromium/net/impl/CronetBidirectionalStream.java

```diff
@@ -233,259 +174,237 @@ public void run() {
+    for (WriteBuffer writeBuffer : mFlushData) {
+      flushData.add(writeBuffer.mByteBuffer.asReadOnlyBuffer());
     }
+    return flushData;
   }
 
   @Override
   public void cancel() {
-    synchronized (mNativeStreamLock) {
-      if (isDoneLocked() || mReadState == State.NOT_STARTED) {
-        return;
-      }
-      mReadState = mWriteState = State.CANCELED;
-      destroyNativeStreamLocked(true);
+    switch (mState.nextAction(Event.USER_CANCEL)) {
+    case NextAction.CANCEL:
+      mStream.cancel();
+      break;
+    case NextAction.PROCESS_CANCEL:
+      onCanceledReceived();
+      break;
+    case NextAction.CARRY_ON:
+    case NextAction.TAKE_NO_MORE_ACTIONS:
+      // Has already been cancelled, an error condition already registered, or just too late.
+      break;
+    default:
+      assert false;
     }
   }
 
   @Override
   public boolean isDone() {
-    synchronized (mNativeStreamLock) { return isDoneLocked(); }
+    return mState.isDone();
   }
 
-  @GuardedBy("mNativeStreamLock")
-  private boolean isDoneLocked() {
-    return mReadState != State.NOT_STARTED && mNativeStream == 0;
+  private void onSucceeded() {
+    postTaskToExecutor(new Runnable() {
+      @Override
+      public void run() {
+        onSucceededOnExecutor();
+      }
+    });
   }
 
   /*
    * Runs an onSucceeded callback if both Read and Write sides are closed.
    */
-  private void maybeOnSucceededOnExecutor() {
-    synchronized (mNativeStreamLock) {
-      if (isDoneLocked()) {
-        return;
-      }
-      if (!(mWriteState == State.WRITING_DONE && mReadState == State.READING_DONE)) {
-        return;
-      }
-      mReadState = mWriteState = State.SUCCESS;
-      // Destroy native stream first, so UrlRequestContext could be shut
-      // down from the listener.
-      destroyNativeStreamLocked(false);
-    }
+  private void onSucceededOnExecutor() {
+    cleanup();
     try {
       mCallback.onSucceeded(CronetBidirectionalStream.this, mResponseInfo);
     } catch (Exception e) {
       Log.e(CronetUrlRequestContext.LOG_TAG, "Exception in onSucceeded method", e);
     }
   }
 
-  @SuppressWarnings("unused")
-  // TODO(carloseltuerto) Hook up Envoy-Mobile to call back this method.
-  private void onStreamReady(final boolean requestHeadersSent) {
+  private void onStreamReady() {
     postTaskToExecutor(new Runnable() {
       @Override
       public void run() {
-        synchronized (mNativeStreamLock) {
-          if (isDoneLocked()) {
-            return;
-          }
-          mRequestHeadersSent = requestHeadersSent;
-          mReadState = State.WAITING_FOR_READ;
-          if (!doesMethodAllowWriteData(mInitialMethod) && mRequestHeadersSent) {
-            mWriteState = State.WRITING_DONE;
-          } else {
-            mWriteState = State.WAITING_FOR_FLUSH;
-          }
-        }
-
         try {
-          mCallback.onStreamReady(CronetBidirectionalStream.this);
+          if (!mState.isTerminating()) {
```

Comment:
With Cronet, yes. For example, if you send and "http" request (without the "s"), then this ends up with this callback. I had to mimic that, because EM does not behave like this.